File: network-equipment-pms/pms-deliverable/src/main/java/com/dp/plat/deliverable/entity/Deliverable.java
package com.dp.plat.deliverable.entity;

import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableName;
import com.dp.plat.common.dto.DeliverableContentBlock;
import com.dp.plat.common.entity.BaseEntity;
import com.dp.plat.common.handler.JsonTypeHandlers;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 交付件实体（全生命周期 7 态状态机）。
 *
 * <p>关联设计文档：§2.2（行 185-196）、§3.4 交付件状态机 7 态（行 393-428）。
 *
 * <p>状态流转：DRAFT → SUBMITTED → REVIEWED → SIGNED → PUBLISHED → REFERENCED → ARCHIVED
 * （其中 SUBMITTED/REVIEWED 可退回 DRAFT；PUBLISHED 可经修订新建版本回到 DRAFT）。
 *
 * <p>注：本实体映射 {@code pms_deliverable} 表（V2 建表、V70 扩展）。
 * pms-project 模块下的旧 {@code com.dp.plat.project.entity.Deliverable} 保留用于历史代码兼容，
 * 本模块为全生命周期的单一权威实体（含 7 态字段 + 版本/签核时间戳）。</p>
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
@TableName(value = "pms_deliverable", autoResultMap = true)
public class Deliverable extends BaseEntity {

    /** 项目ID。 */
    private Long projectId;

    /** 交付件名称。 */
    private String deliverableName;

    /** 交付件性质类型（见字典 pms_deliverable_type）：DOCUMENT/CODE/ENTITY_REF/MODEL/CONFIG/DATA/OTHER。 */
    private String deliverableType;

    /** 文件路径（最新版本）。 */
    private String filePath;

    /**
     * 结构化内容块（JSON 数组，元素见 {@link DeliverableContentBlock}）。
     *
     * <p>借鉴问卷功能：交付件文档内容由若干内容块有序组成，支持富文本/内嵌表/选项卡/
     * 标题/分隔线/代码块 6 种类型（blockType 取值由字典 {@code pms_deliverable_block_type} 维护）。
     * 新建交付件时由前端按 deliverableType 加载对应默认模板
     * （{@link DeliverableTypeTemplate}）填充，用户可二次编辑。</p>
     *
     * <p>TypeHandler：{@link JsonTypeHandlers.DeliverableContentBlockListHandler}
     * （重写 parse 以 {@code TypeReference<List<DeliverableContentBlock>>} 反序列化，
     * 解决泛型擦除）。{@code autoResultMap = true} 已在类级 {@code @TableName} 开启，
     * 确保 BaseMapper 查询时字段级 typeHandler 生效。</p>
     */
    @TableField(typeHandler = JsonTypeHandlers.DeliverableContentBlockListHandler.class)
    private List<DeliverableContentBlock> contentBlocks;

    /**
     * 状态（7 态）：DRAFT/SUBMITTED/REVIEWED/SIGNED/PUBLISHED/REFERENCED/ARCHIVED。
     *
     * <p>对应 {@link com.dp.plat.deliverable.enums.DeliverableStatus}。</p>
     */
    private String status;

    /** 所属阶段ID。 */
    private Long phaseId;

    /** 当前版本号，从 1 开始。修订时 +1 并新建 {@link DeliverableVersion} 记录。 */
    private Integer currentVersion;

    /** 是否必需交付件（影响阶段退出校验）。 */
    private Boolean mandatory;

    /** 是否模板预设（模板实例化创建 = true，过程新增 = false）。 */
    private Boolean templateInherited;

    /** 签核角色。 */
    private String approverRole;

    /** 引用实体类型（见字典 pms_deliverable_ref_entity_type：TASK/ASSET/PHASE/PROJECT/DELIVERABLE/REPORT）。 */
    private String refEntityType;

    /** 引用实体ID（当 deliverableType=ENTITY_REF 时指向具体业务对象）。 */
    private Long refEntityId;

    /** 发布时间（SIGNED → PUBLISHED 时写入）。 */
    private LocalDateTime publishedAt;

    /** 归档时间（REFERENCED → ARCHIVED 时写入）。 */
    private LocalDateTime archivedAt;
}

File: network-equipment-pms/pms-frontend/src/validators/deliverable.ts
/**
 * Deliverable 实体 Schema 与 Validator
 * ===========================================================================
 *
 * 与后端 `pms-deliverable/.../entity/Deliverable.java` 严格对齐，作为后端
 * 数据契约的前端镜像。
 *
 * 注：后端 Deliverable 实体未声明 Jakarta Validation 注解（无 @NotBlank/@NotNull
 * 等约束），但本 schema 仍按字段一一对应建立规则，用于：
 * 1. 出站请求体字段白名单过滤（防止前端误传未定义字段污染后端）；
 * 2. 入站响应数据字段白名单过滤（防止后端多余字段污染前端状态）。
 *
 * 字段命名以后端为准：projectId / deliverableName / deliverableType / status 等。
 *
 * 历史问题：前端 `api/deliverable.ts` 的 `Deliverable` interface 字段名已与
 * 后端基本一致，故 `fieldMapping` 为空对象，仅做白名单过滤。
 * ===========================================================================
 */
import {
  createValidatorWithMapping,
  defineSchema,
  number,
  optional,
  string,
  boolean,
  type Schema,
  type ValidationResult
} from './index'
import { registerValidator } from './registry'

// ============ Deliverable Schema ============
//
// 与后端 Deliverable.java 字段一一对应（实体本身无 Jakarta Validation 注解）：
// - projectId:        Long
// - deliverableName:  String
// - deliverableType:  String [DOCUMENT/CONFIG/REPORT/OTHER]
// - filePath:         String
// - status:           String [DRAFT/SUBMITTED/REVIEWED/SIGNED/PUBLISHED/REFERENCED/ARCHIVED]
// - phaseId:          Long
// - currentVersion:   Integer
// - mandatory:        Boolean
// - approverRole:     String
// - publishedAt:      LocalDateTime
// - archivedAt:       LocalDateTime
//
// 注：BaseEntity 字段（id/createTime/updateTime/createBy/updateBy）
//     在前端 schema 中以可选形式声明，便于响应数据校验。

const DELIVERABLE_TYPE_ENUM = ['DOCUMENT', 'CONFIG', 'REPORT', 'OTHER'] as const

const DELIVERABLE_STATUS_ENUM = [
  'DRAFT',
  'SUBMITTED',
  'REVIEWED',
  'SIGNED',
  'PUBLISHED',
  'REFERENCED',
  'ARCHIVED'
] as const

export const deliverableSchema: Schema = defineSchema({
  // ============ BaseEntity 字段 ============
  id: [optional(), number()],
  createTime: [optional(), string()],
  updateTime: [optional(), string()],
  createBy: [optional(), string()],
  updateBy: [optional(), string()],

  // ============ Deliverable 自身字段 ============
  projectId: [optional(), number()],
  deliverableName: [optional(), string()],
  deliverableType: [optional(), string()],
  filePath: [optional(), string()],
  status: [optional(), string()],
  phaseId: [optional(), number()],
  currentVersion: [optional(), number()],
  mandatory: [optional(), boolean()],
  approverRole: [optional(), string()],
  publishedAt: [optional(), string()],
  archivedAt: [optional(), string()]
})

/**
 * 旧字段名 → 后端字段名映射。
 *
 * 前端 `api/deliverable.ts` 的 `Deliverable` interface 字段名已与后端一致，
 * 无需映射，置为空对象。
 */
export const deliverableFieldMapping = {} as const

/** Deliverable 请求体 validator（字段名已一致，无映射） */
export const deliverableRequestValidator = createValidatorWithMapping(
  deliverableSchema,
  deliverableFieldMapping
)

/** Deliverable 响应数据 validator（无映射，直接按后端字段名校验） */
export function validateDeliverableResponse(input: unknown): ValidationResult {
  return createValidatorWithMapping(deliverableSchema, {})(input)
}

/**
 * Deliverable 类型定义——与后端 Deliverable 实体字段名严格一致。
 *
 * 替代 `api/deliverable.ts` 中的旧 `Deliverable` interface。
 * 现有 view 代码迁移到新 interface 后，可删除旧定义。
 */
export interface DeliverableDTO {
  // BaseEntity
  id?: number
  createTime?: string
  updateTime?: string
  createBy?: string
  updateBy?: string

  // Deliverable 自身字段
  projectId?: number
  deliverableName?: string
  deliverableType?: typeof DELIVERABLE_TYPE_ENUM[number]
  filePath?: string
  status?: typeof DELIVERABLE_STATUS_ENUM[number]
  phaseId?: number
  currentVersion?: number
  mandatory?: boolean
  approverRole?: string
  publishedAt?: string
  archivedAt?: string
}

// ============ 注册到全局 registry ============
//
// 注册后，request.ts 拦截器会自动调用 deliverableRequestValidator 校验以下 URL：
// - POST /api/deliverable      （创建交付件）
// - PUT  /api/deliverable/{id} （更新交付件基础信息）
//
// 校验失败 → ElMessage.error + 阻止请求发送
registerValidator({
  method: 'POST',
  urlPattern: '/api/deliverable',
  requestValidator: deliverableRequestValidator,
  description: '创建交付件 - 校验请求体字段名'
})

registerValidator({
  method: 'PUT',
  urlPattern: /^\/api\/deliverable\/\d+$/,
  requestValidator: deliverableRequestValidator,
  description: '更新交付件 - 校验请求体字段名'
})
